pkg/reconciler/workspace/controller.go

```diff
@@ -0,0 +1,318 @@
+}
+
+type Controller struct {
+	queue workqueue.RateLimitingInterface
+
+	kcpClient        kcpclient.ClusterInterface
+	workspaceIndexer cache.Indexer
+	workspaceLister  tenancylister.WorkspaceLister
+
+	workspaceShardIndexer cache.Indexer
+	workspaceShardLister  tenancylister.WorkspaceShardLister
+}
+
+func (c *Controller) enqueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	klog.Infof("queueing workspace %q", key)
+	c.queue.Add(key)
+}
+
+func (c *Controller) enqueueAddedShard(obj interface{}) {
+	shard, ok := obj.(*tenancyv1alpha1.WorkspaceShard)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("got %T when handling added WorkspaceShard", obj))
+		return
+	}
+	klog.Infof("handling added shard %q", shard.Name)
+	workspaces, err := c.workspaceIndexer.ByIndex(unschedulable, "true")
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	for _, workspace := range workspaces {
+		key, err := cache.MetaNamespaceKeyFunc(workspace)
+		if err != nil {
+			runtime.HandleError(err)
+			return
+		}
+		klog.Infof("queuing unschedulable workspace %q", key)
+		c.queue.Add(key)
+	}
+}
+
+func (c *Controller) enqueueDeletedShard(obj interface{}) {
+	shard, ok := obj.(*tenancyv1alpha1.WorkspaceShard)
+	if !ok {
+		runtime.HandleError(fmt.Errorf("got %T when handling deleted WorkspaceShard", obj))
+		return
+	}
+	klog.Infof("handling removed shard %q", shard.Name)
+	workspaces, err := c.workspaceIndexer.ByIndex(targetShardName, shard.Name)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	for _, workspace := range workspaces {
+		key, err := cache.MetaNamespaceKeyFunc(workspace)
+		if err != nil {
+			runtime.HandleError(err)
+			return
+		}
+		klog.Infof("queuing orphaned workspace %q", key)
+		c.queue.Add(key)
+	}
+}
+
+func (c *Controller) Start(ctx context.Context, numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Info("Starting Workspace controller")
+	defer klog.Info("Shutting down Workspace controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+	}
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	klog.Infof("processing key %q", key)
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	err := c.process(ctx, key)
+	c.handleErr(err, key)
+	return true
+}
+
+func (c *Controller) handleErr(err error, key string) {
+	// Reconcile worked, nothing else to do for this workqueue item.
+	if err == nil {
+		klog.Infof("Successfully reconciled %q", key)
+		c.queue.Forget(key)
+		return
+	}
+
+	if errors.IsRetryable(err) {
```

Comment:
compare https://github.com/kcp-dev/kcp/pull/219/files#r752915735. I think this logic is wrong (and kind of dangerous if we continue this and spread the pattern).